cmd/release-controller-api/http.go

```diff
@@ -153,12 +163,272 @@ func (c *Controller) userInterfaceHandler() http.Handler {
+		epicWithoutFeatureMap[epicDetails.Epic] = append(epicWithoutFeatureMap[epicDetails.Parent], child)
+	}
+
+	for epic, children := range epicWithoutFeatureMap {
+		f := &FeatureTree{
+			IssueKey:        epic,
+			Summary:         mapIssueDetails[epic].Summary,
+			Description:     mapIssueDetails[epic].Description,
+			ReleaseNotes:    mapIssueDetails[epic].ReleaseNotes,
+			Type:            mapIssueDetails[epic].IssueType,
+			Epic:            mapIssueDetails[epic].Epic,
+			Feature:         mapIssueDetails[epic].Feature,
+			Parent:          mapIssueDetails[epic].Parent,
+			NotLinkedType:   sectionTypeNoFeatureWithEpic,
+			ResolutionDate:  mapIssueDetails[epic].ResolutionDate,
+			PRs:             mapIssueDetails[epic].PRs,
+			IncludedOnBuild: true,
+			Children:        children,
+		}
+		featureTrees = append(featureTrees, f)
+	}
+
+	// TODO - find a better way to do this, this it is to expensive
+	redistributedUnknowns := sets.String{}
+	for _, unknown := range unknowns {
+		if unknown.Parent != "" {
+			redistributeUnknowns(featureTrees, unknown.Parent, unknown, &redistributedUnknowns)
+		}
+		if unknown.Epic != "" {
+			redistributeUnknowns(featureTrees, unknown.Epic, unknown, &redistributedUnknowns)
+		}
+		if unknown.Feature != "" {
+			redistributeUnknowns(featureTrees, unknown.Feature, unknown, &redistributedUnknowns)
+		}
+	}
+	for _, ticket := range unknowns {
+		if !redistributedUnknowns.Has(ticket.IssueKey) {
+			ticket.NotLinkedType = sectionTypeUnsortedUnknowns
+			featureTrees = append(featureTrees, ticket)
+		}
+	}
+
+	// remove every tree from sectionTypeNoEpicNoFeature that has no PRs, since it means that it is not part of the
+	// change log, i.e. Cards part of the parent/epics/features group gathered for the featureTree, but not linked
+	// properly (e.g. an Epic that links directly to a "Market Problem" instead of a Feature, and Feature is the root)
+	for _, ticket := range featureTrees {
+		if ticket.NotLinkedType == sectionTypeNoEpicNoFeature {
+			if isPRsEmpty(ticket) {
+				removeFeatureTree(featureTrees, ticket.IssueKey)
+			}
+		}
+	}
+
+	return featureTrees, nil
+}
+
+func removeFeatureTree(slice []*FeatureTree, issueKey string) {
+	for i, ft := range slice {
+		if ft.IssueKey == issueKey {
+			// Remove the element by swapping it with the last element
+			slice[i] = slice[len(slice)-1]
+			slice = slice[:len(slice)-1]
+			break
+		}
+	}
+}
+
+func isPRsEmpty(ft *FeatureTree) bool {
+	if len(ft.PRs) > 0 {
+		return false
+	}
+	for _, child := range ft.Children {
+		if !isPRsEmpty(child) {
+			return false
+		}
+	}
+	return true
+}
+
+func redistributeUnknowns(slice []*FeatureTree, key string, feature *FeatureTree, s *sets.String) bool {
+	for _, node := range slice {
+		if node.IssueKey == key {
+			node.Children = append(node.Children, feature)
+			s.Insert(feature.IssueKey)
+			return true
+		}
+		redistributeUnknowns(node.Children, key, feature, s)
+	}
+	return false
+}
```

Comment:
You put cycle limits on your other recursive functions, should this have one as well?

Reply:
Added the cycle limit